File: Leetcode_By_Ashika/0076-minimum-window-substring/solution.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if (s.empty() || t.empty()) return "";

        int total = t.length();
        unordered_map<char, int> map;

        // store freq of t
        for (char c : t) {
            map[c]++;
        }

        int start = 0, end = 0;
        int minLen = INT_MAX, minStart = 0;

        while (end < s.length()) {
            char cEnd = s[end];
            if (map.find(cEnd) != map.end()) {
                map[cEnd]--;
                if (map[cEnd] >= 0) {
                    total--;
                }
            }

            // shrink window from left if all chars found
            while (total == 0) {
                if (end - start + 1 < minLen) {
                    minLen = end - start + 1;
                    minStart = start;
                }

                char cStart = s[start];
                if (map.find(cStart) != map.end()) {
                    map[cStart]++;
                    if (map[cStart] > 0) {
                        total++;
                    }
                }
                start++;
            }
            end++;
        }

        return (minLen == INT_MAX) ? "" : s.substr(minStart, minLen);
    }
};

File: Leetcode_By_Ashika/0076-minimum-window-substring/solution.java
class Solution {
    public String minWindow(String s, String t) {
        if (s.length() == 0 || t.length() == 0) return "";

        int[] freq = new int[128];
        for (char c : t.toCharArray()) {freq[c]++;}

        int start = 0, end = 0;
        int total = t.length();
        int minLen = Integer.MAX_VALUE, minStart = 0;

        while (end < s.length()) {
            char ch = s.charAt(end);
            if (freq[ch] > 0) 
            total--;  // char is needed
            freq[ch]--;
            end++;  // expand window

            // shrink window while valid
            while (total == 0) {
                if (end - start < minLen) {
                    minLen = end - start;
                    minStart = start;
                }

                char startChar = s.charAt(start);
                freq[startChar]++;
                if (freq[startChar] > 0) total++;  // window lost a required char
                start++;  // shrink
            }
        }

        return (minLen == Integer.MAX_VALUE) ? "" : s.substring(minStart, minStart + minLen);
    }
}

